concurrency/IQRouter.java

```diff
@@ -0,0 +1,708 @@
+                }
+            }
+        }
+        catch (Exception e) {
+            Log.error(LocaleUtils.getLocalizedString("admin.error.routing"), e);
+            Session session = sessionManager.getSession(packet.getFrom());
+            if (session != null) {
+                IQ reply = IQ.createResultIQ(packet);
+                reply.setError(PacketError.Condition.internal_server_error);
+                session.process(reply);
+            }
+            /*** BEGIN amazon ***/
+            isSuccessful = false;
+            if (metric != null) {
+                metric.addCount((session != null ? METRIC_ROUTE_MESSAGE_FAILURE_INTERNAL_SERVER_ERROR :
+                        METRIC_ROUTE_MESSAGE_FAILURE_SESSION_NULL), 1, Unit.ONE);
+            }
+            /*** END amazon ***/
+        }
+        return isSuccessful; /*** SINGLE LINE amazon ***/
+    }
+
+    private void sendErrorPacket(IQ originalPacket, PacketError.Condition condition, final Metrics
+            metric) { /*** SINGLE LINE amazon ***/
+        if (IQ.Type.error == originalPacket.getType()) {
+            /*** BEGIN amazon ***/
+            if (metric != null) {
+                metric.addCount(METRIC_ROUTE_MESSAGE_FAILURE_RESPONSE_TO_IQ_ERROR, 1, Unit.ONE);
+            }
+            /*** END amazon ***/
+            Log.error("Cannot reply an IQ error to another IQ error: " + originalPacket.toXML());
+            return;
+        }
+        if (originalPacket.getFrom() == null) {
+            if (Log.isDebugEnabled()) {
+                Log.debug("Original IQ has no sender for reply; dropped: " + originalPacket.toXML());
+            }
+            return;
+        }
+        IQ reply = IQ.createResultIQ(originalPacket);
+        reply.setChildElement(originalPacket.getChildElement().createCopy());
+        reply.setError(condition);
+        /*** BEGIN amazon ***/
+        if (metric != null) {
+            metric.addCount(METRIC_FAILURE_PREFIX + condition.toString(), 1, Unit.ONE);
+        }
+        /*** END amazon ***/
+        // Check if the server was the sender of the IQ
+        if (serverName.equals(originalPacket.getFrom().toString())) {
+            // Just let the IQ router process the IQ error reply
+            handle(reply, metric); /*** SINGLE LINE amazon ***/
+            return;
+        }
+        // Route the error packet to the original sender of the IQ.
+        routingTable.routePacket(reply.getTo(), reply, true);
+    }
+
+    private IQHandler getHandler(String namespace) {
+        IQHandler handler = namespace2Handlers.get(namespace);
+        if (handler == null) {
+            for (IQHandler handlerCandidate : iqHandlers) {
+                IQHandlerInfo handlerInfo = handlerCandidate.getInfo();
+                if (handlerInfo != null && namespace.equalsIgnoreCase(handlerInfo.getNamespace())) {
+                    handler = handlerCandidate;
+                    namespace2Handlers.put(namespace, handler);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **576, 577, and 582**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **587**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.